adapter/kubernetes/cache.go

```diff
@@ -211,3 +298,76 @@ func (e eventType) String() string {
 		return "Unknown"
 	}
 }
+
+func (c *controllerImpl) updatePodSvcMap(obj interface{}) {
+	svc, ok := obj.(*v1.Service)
+	if !ok {
+		c.env.Logger().Warningf("received update for non-service item")
+		return
+	}
+	labelSelectorMap := svc.Spec.Selector
+	listOpts := metav1.ListOptions{LabelSelector: labels.FormatLabels(labelSelectorMap)}
+	list, err := c.clientset.CoreV1().Pods(svc.Namespace).List(listOpts)
+	if err != nil {
+		c.env.Logger().Warningf("Could not list pods for service '%s': %v, skipping", svc.Name, err)
+		return
+	}
+	c.podSvcMapMutex.Lock()
+	for _, pod := range list.Items {
+		c.podSvcMap[key(pod.Namespace, pod.Name)] = key(svc.Namespace, svc.Name)
+	}
+	c.podSvcMapMutex.Unlock()
+}
+
+func (c *controllerImpl) deleteFromPodSvcMap(obj interface{}) {
+	svc, ok := obj.(*v1.Service)
+	if !ok {
+		c.env.Logger().Warningf("received delete for non-service item")
+		return
+	}
+	toDelete := make([]string, 0, 5)
+	c.podSvcMapMutex.RLock()
+	for podKey, svcKey := range c.podSvcMap {
+		if svcKey == key(svc.Namespace, svc.Name) {
+			toDelete = append(toDelete, podKey)
+		}
+	}
+	c.podSvcMapMutex.RUnlock()
+	c.podSvcMapMutex.Lock()
+	for _, deleteKey := range toDelete {
+		delete(c.podSvcMap, deleteKey)
+	}
+	c.podSvcMapMutex.Unlock()
+}
+
+func (c *controllerImpl) updateIPPodMap(obj interface{}) {
```

Comment:
Say POD p1 belongs to service1 and service2.
service2 changed last, so in our map   p1 --> service2 is the mapping.

Now labels on POD p1 change, so it is no longer in service2.
We will get 2 calls I presume, 1 that the POD has changed. (this one)
And service membership has changed.

If we follow the service membership code, I am not sure it will be removed. 
Deletion only happens when service is deleted. either that or I am missing something.